Use system python packages with --enable-system-site-packages

orlitzky opened the pull request:
Rebased branch of https://github.com/sagemath/sage/issues/29665

Platform-specific testing:
- for example TARGETS_PRE="build-local" tox -e docker-ubuntu-focal-standard-sitepackages -- build ptest

Fixes #29665

orlitzky committed:
Trac #29665: add --enable-system-site-packages option to ./configure.

This new option is experimental and disabled by default. When it is
enabled, the string "--system-site-packages" is appended to a new
environment variable SAGE_VENV_FLAGS that is then passed to the
sage-venv command during the build using autoconf substitution.
Thusly the correct setting finds its way into local/pyvenv.cfg.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for py.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for wheel.

orlitzky committed:
Trac #29665: new spkg-configure.m4 for cvxopt.

This includes a SAGE_SPKG_DEPCHECK([gsl glpk suitesparse],... because
those C dependencies wind up linked into shared libraries like,

  /usr/.../cvxopt/glpk.cpython-39-x86_64-linux-gnu.so

that could potentially conflict with (for example) the glpk SPKG.

orlitzky committed:
Trac #29665: new spkg-configure.m4 for pyzmq.

This includes a SAGE_SPKG_DEPCHECK([zeromq],... because libzmq winds
up linked into shared libraries like,

  /usr/.../zmq/backend/cython/message.cpython-39-x86_64-linux-gnu.so

that potentially conflict with the zeromq SPKG.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for requests.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for six.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for packaging.

orlitzky committed:
Trac #29665: new spkg-configure.m4 for numpy.

This includes a SAGE_SPKG_DEPCHECK([openblas],... because blas/lapack
wind up linked into shared libraries like,

  /usr/.../numpy/linalg/lapack_lite.cpython-39-x86_64-linux-gnu.so

that potentially conflict with the openblas SPKG.

orlitzky committed:
Trac #29665: new spkg-configure.m4 for scipy.

This includes a SAGE_SPKG_DEPCHECK([openblas],... because blas/lapack
wind up linked into shared libraries like,

  /usr/.../scipy/linalg/_flinalg.cpython-39-x86_64-linux-gnu.so

that potentially conflict with the openblas SPKG.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pluggy.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pickleshare.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for html5lib.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for networkx.

orlitzky committed:
Trac #29665: new SAGE_PYTHON_PACKAGE_CHECK() autoconf macro.

This new macro encapsulates the check for system installations of
python packages. It uses setuptools.version's pkg_resources to
check if the contents of install-requires.txt are satisfied for
the given package; and if not, tells the build system to install
the SPKG instead.

orlitzky committed:
Trac #29665: new spkg-configure.m4 for matplotlib.

This includes a SAGE_SPKG_DEPCHECK([bzip2 freetype libpng qhull],...
because those dependencies wind up linked into shared libraries like,

  /usr/lib/.../matplotlib/ft2font.cpython-39-x86_64-linux-gnu.so

that potentially conflict with their SPKG counterparts.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for bleach.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for decorator.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for idna.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for certifi.

orlitzky committed:
Trac #29665: new spkg-configure.m4 for pillow.

This includes an SAGE_SPKG_DEPCHECK([bzip2 freetype libpng zlib],...
because those dependencies wind up linked into shared libraries like,

  /usr/.../PIL/_imagingft.cpython-39-x86_64-linux-gnu.so

and potentially conflict with their SPKG counterparts.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for mpmath.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for sympy.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pygments.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for jinja2.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pandocfilters.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for webencodings.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pytz.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for texttable.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for babel.

orlitzky committed:
Trac #29665: force system python3 with --enable-system-site-packages.

When the python SPKG is used, the --enable-system-site-packages flag
will essentially do nothing, and that may confuse users who have
wasted a few hours building sage. (Generally, they will want to fix
detection of their system python3 first.)

To forestall that, we now set "--with-system-python3=force" whenever
the "--enable-system-site-packages" flag is passed.

orlitzky committed:
Trac #29665: poison PYTHONUSERBASE when testing for "system" packages.

When we're detecting system installations of python packages, we want
to hide anything that was installed with e.g. "pip install --user". To
accomplish that, we mimic the PYTHONUSERBASE poisoning from sage-env
within the SAGE_PYTHON_PACKAGE_CHECK() macro.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for setuptools.

orlitzky committed:
Trac #29665: run SAGE_PYTHON_PACKAGE_CHECK() inside a venv.

To recreate the runtime environment as closely as possible, we now use
a venv (with system site packages enabled) when checking for system
copies of python packages.

orlitzky committed:
Trac #29665: log SAGE_PYTHON_PACKAGE_CHECK() failures to config.log.

orlitzky committed:
Trac #29665: don't clobber a plain file named config.venv.

When SAGE_PYTHON_PACKAGE_CHECK() is cleaning up, we no longer remove
"config.venv" unconditionally. Instead, we remove it only if it is a
directory. This is also how "pyvenv --clear" works.

orlitzky committed:
Trac #29665: don't announce system python packages that won't be used.

When --enable-system-site-packages was not given, the python packages
using SAGE_PYTHON_PACKAGE_CHECK() in spkg-configure.m4 cannot be used.
Rather than suggest that users install them anyway, that macro now
hacks the value of sage_use_system_$package to be "no" when the use
of system site packages is disabled at the python level.

orlitzky committed:
Trac #29665: add Gentoo package information for python packages.

Commit:
build/pkgs: Add distros/fedora.txt for Python packages

Commit:
build/pkgs: Add distros/arch.txt for Python packages

Commit:
build/pkgs: Add distros/opensuse.txt for Python packages

Commit:
build/pkgs: Add distros/debian.txt for Python packages

orlitzky committed:
Trac #29665: improve SAGE_PYTHON_PACKAGE_CHECK() failure message.

When --enable-system-site-packages is off, SAGE_PYTHON_PACKAGE_CHECK()
will always install the corresponding SPKG. We improve the ./configure
output in that case to make it clear that the decision was based on
the value of --enable-system-site-packages and not the "real" check.

orlitzky committed:
Trac #29665: add more documentation for SAGE_PYTHON_PACKAGE_CHECK().

An earlier commit allows SAGE_PYTHON_PACKAGE_CHECK(package) to tweak the
value of sage_use_system_$package if --enable-system-site-packages was
not used. This essentially changes the default --with-system-$package=yes
to "no" in that case. Here we update the documentation to reflect it.

orlitzky committed:
Trac #29665: use sage-venv in SAGE_PYTHON_PACKAGE_CHECK().

For consistency with the rest of the build system, we now use the
build/bin/sage-venv script to create the venv in this autoconf macro
rather than invoking $PYTHON_FOR_VENV with "-m venv" directly.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for urllib3.

orlitzky committed:
Trac #29665: add Gentoo package information for urllib3.

orlitzky committed:
Trac #29665: add Gentoo package information for nbformat.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for nbformat.

orlitzky committed:
Trac #29665: add Gentoo package information for zipp.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for zipp.

orlitzky committed:
Trac #29665: add Gentoo package information for wcwidth.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for wcwidth.

orlitzky committed:
Trac #29665: add Gentoo package information for jedi.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for jedi.

orlitzky committed:
Trac #29665: add Gentoo package information for tomli.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for tomli.

orlitzky committed:
Trac #29665: add Gentoo package information for defusedxml.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for defusedxml.

orlitzky committed:
Trac #29665: add Gentoo package information for ipywidgets.

orlitzky committed:
Trac #29665: add Gentoo package information for simplegeneric.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for simplegeneric.

orlitzky committed:
Trac #29665: add Gentoo package information for imagesize.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for imagesize.

orlitzky committed:
Trac #29665: add Gentoo package information for mistune.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for mistune.

orlitzky committed:
Trac #29665: add Gentoo package information for backcall.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for backcall.

orlitzky committed:
Trac #29665: add Gentoo package information for parso.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for parso.

orlitzky committed:
Trac #29665: add Gentoo package information for pyparsing.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pyparsing.

orlitzky committed:
Trac #29665: add Gentoo package information for debugpy.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for debugpy.

orlitzky committed:
Trac #29665: add Gentoo package information for tzlocal.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for tzlocal.

orlitzky committed:
Trac #29665: add Gentoo package information for pycparser.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pycparser.

orlitzky committed:
Trac #29665: add Gentoo package information for cycler.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for cycler.

orlitzky committed:
Trac #29665: add Gentoo package information for traitlets.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for traitlets.

orlitzky committed:
Trac #29665: add Gentoo package information for ptyprocess.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for ptyprocess.

orlitzky committed:
Trac #29665: add Gentoo package information for attrs.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for attrs.

orlitzky committed:
Trac #29665: add Gentoo package information for send2trash.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for send2trash.

orlitzky committed:
Trac #29665: add Gentoo package information for pyrsistent.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pyrsistent.

orlitzky committed:
Trac #29665: add Gentoo package information for entrypoints.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for entrypoints.

orlitzky committed:
Trac #29665: add Gentoo package information for markupsafe.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for markupsafe.

orlitzky committed:
Trac #29665: add Gentoo package information for flit_core.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for flit_core.

orlitzky committed:
Trac #29665: add Gentoo package information for nest_asyncio.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for nest_asyncio.

orlitzky committed:
Trac #29665: add Gentoo package information for ipython_genutils.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for ipython_genutils.

orlitzky committed:
Trac #29665: add Gentoo package information for ipython.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for ipython.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for ipywidgets.

orlitzky committed:
Trac #29665: add Gentoo package information for wheel.

orlitzky committed:
Trac #29665: set $SAGE_PIP_INSTALL_FLAGS for --enable-system-site-packages.

We need to pass some special flags to "pip install" to make it do the
right thing in the precence of system-site packages. Here we introduce
a new autoconf variable, SAGE_PIP_INSTALL_FLAGS, that will be used
to pass those flags to the build system.

With autoconf aware of the variable, we substitute its value into the
sage-build-env-config script for use in spkg-install.in.

orlitzky committed:
Trac #29665: pass $SAGE_PIP_INSTALL_FLAGS to sdh_pip_install.

Since sdh_pip_install is usually called from spkg-install.in, the
variables from sage-build-env-config will be available. Here we pass
the contents of $SAGE_PIP_INSTALL_FLAGS to sage-pip-install
unconditionally, and hope that the user ignores this variable unless
he truly knows what he is doing.

orlitzky committed:
Trac #29665: add Gentoo package information for terminado.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for terminado.

orlitzky committed:
Trac #29665: add Gentoo package information for cffi.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for cffi.

orlitzky committed:
Trac #29665: add Gentoo package information for pip.

orlitzky committed:
Trac #29665: new spkg-configure.m4 for pip.

orlitzky committed:
Trac #29665: add Gentoo package information for pexpect.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pexpect.

orlitzky committed:
Trac #29665: add Gentoo package information for docutils.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for docutils.

orlitzky committed:
Trac #29665: add Gentoo package information for nbclient.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for nbclient.

orlitzky committed:
Trac #29665: add Gentoo package information for argon2_cffi.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for argon2_cffi.

orlitzky committed:
Trac #29665: fix package name in dateutil's install-requires.txt.

orlitzky committed:
Trac #29665: add Gentoo package information for dateutil.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for dateutil.

orlitzky committed:
Trac #29665: add Gentoo package information for setuptools_scm.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for setuptools_scm.

orlitzky committed:
Trac #29665: add Gentoo package information for nbconvert.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for nbconvert.

orlitzky committed:
Trac #29665: add Gentoo package information for jupyter_core.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for jupyter_core.

orlitzky committed:
Trac #29665: add Gentoo package information for jupyter_client.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for jupyter_client.

orlitzky committed:
Trac #29665: add Gentoo package information for snowballstemmer.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for snowballstemmer.

orlitzky committed:
Trac #29665: add Gentoo package information for tornado.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for tornado.

orlitzky committed:
Trac #29665: add Gentoo package information for jsonschema.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for jsonschema.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for widgetsnbextension.

orlitzky committed:
Trac #29665: add Gentoo package information for importlib_resources.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for importlib_resources.

orlitzky committed:
Trac #29665: add Gentoo package information for kiwisolver.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for kiwisolver.

orlitzky committed:
Trac #29665: add Gentoo package information for notebook.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for notebook.

orlitzky committed:
Trac #29665: add Gentoo package information for matplotlib_inline.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for matplotlib_inline.

orlitzky committed:
Trac #29665: add Gentoo package information for ipykernel.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for ipykernel.

orlitzky committed:
Trac #29665: add Gentoo package information for prompt_toolkit.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for prompt_toolkit.

orlitzky committed:
Trac #29665: add Gentoo package information for charset_normalizer.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for charset_normalizer.

orlitzky committed:
Trac #29665: add Gentoo package information for prometheus_client.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for prometheus_client.

orlitzky committed:
Trac #29665: add Gentoo package information for pybind11.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pybind11.

orlitzky committed:
Trac #29665: add Gentoo package information for pkgconfig.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pkgconfig.

orlitzky committed:
Trac #29665: add Gentoo package information for sphinxcontrib_serializinghtml.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for sphinxcontrib_serializinghtml.

orlitzky committed:
Trac #29665: add Gentoo package information for sphinxcontrib_jsmath.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for sphinxcontrib_jsmath.

orlitzky committed:
Trac #29665: add Gentoo package information for sphinxcontrib_devhelp.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for sphinxcontrib_devhelp.

orlitzky committed:
Trac #29665: add Gentoo package information for sphinxcontrib_qthelp.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for sphinxcontrib_qthelp.

orlitzky committed:
Trac #29665: add Gentoo package information for sphinxcontrib_applehelp.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for sphinxcontrib_applehelp.

orlitzky committed:
Trac #29665: add Gentoo package information for sphinxcontrib_htmlhelp.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for sphinxcontrib_htmlhelp.

orlitzky committed:
Trac #29665: add Gentoo package information for jupyterlab_pygments.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for jupyterlab_pygments.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for ply.

orlitzky committed:
Trac #29665: add Gentoo package information for ply.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for gast.

orlitzky committed:
Trac #29665: add Gentoo package information for gast.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for pythran.

orlitzky committed:
Trac #29665: add Gentoo package information for pythran.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for beniget.

orlitzky committed:
Trac #29665: add Gentoo package information for beniget.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for alabaster.

orlitzky committed:
Trac #29665: add Gentoo package information for alabaster.

orlitzky committed:
Trac #29665: standard python spkg-configure.m4 for importlib_metadata.

orlitzky committed:
Trac #29665: add Gentoo package information for importlib_metadata.

Commit:
build/pkgs/pythran: Update to 0.13.1

Commit:
build/pkgs/meson_python: Update to 0.13.2

Commit:
build/pkgs/pybind11: Update to 2.11.1

Commit:
build/pkgs/pyproject_metadata: Update to 0.7.1

Commit:
build/pkgs/gast: Update to 0.5.4

Commit:
build/pkgs/meson_python/patches: Remove

Commit:
build/pkgs/meson: Update to 1.2.1

Commit:
build/pkgs/ninja_build: Update to 1.11.1

Commit:
build/pkgs/cmake: Update to 3.27.3

Commit:
build/pkgs/scipy/install-requires.txt: Allow 1.11.x

Commit:
build/pkgs/meson/spkg-configure.m4: Require meson >= 1.1.0

Commit:
Merge branch 'cython-0.29.33' into scipy_1_11

orlitzky committed:
m4/sage_python_package_check.m4: update pkg_resources import

The pkg_resources module is now imported with,

  import pkg_resources

instead of

  import setuptools.version.pkg_resources

orlitzky committed:
build/pkgs/pip/distros/gentoo.txt: drop USE=vanilla requirement

orlitzky committed:
build/pkgs/ipython/install-requires.txt: reject <ipython-8.11

Sage isn't ready for ipython-8.11 yet because it requires a newer
prompt_toolkit (https://github.com/sagemath/sage/issues/33428).

mkoeppe committed:
build/pkgs/jupyter_jsmol: Switch to wheel package

orlitzky committed:
configure.ac: set/subst a variable for --enable-system-site-packages

We now set ENABLE_SYSTEM_SITE_PACKAGES=yes and call AC_SUBST on it
when --enable-system-site-packages was passed to ./configure. This
will let us reference it in the build scripts that list available
system packages.

orlitzky committed:
build/bin/sage-build-env-config.in: export ENABLE_SYSTEM_SITE_PACKAGES

This variable is set in configure.ac when --enable-system-site-packages
was passed to ./configure. Having it in the build config allows us to
prune the list of system packages based on that ./configure flag.

orlitzky committed:
build/bin/sage-get-system-packages: optionally skip python packages

Packages calling SAGE_PYTHON_PACKAGE_CHECK in spkg-configure.m4 are
python packages whose detection is enabled by passing the
--enable-system-site-packages flag to ./configure. But when the flag
is NOT passed, we don't want to list the corresponding packages in the
output of sage-get-system-packages. This commit checks for the
presence of such an spkg-configure.m4 file and filters those packages
according to the ENABLE_SYSTEM_SITE_PACKAGES variable.

orlitzky committed:
build/bin/write-dockerfile.sh: refactor using sage-get-system-packages

This avoids duplicating the check for ENABLE_SYSTEM_SITE_PACKAGES.

mkoeppe commented on build/bin/write-dockerfile.sh:
```diff
@@ -15,17 +15,14 @@ SAGE_ROOT=.
+    if [ -n "${SYSTEM_PACKAGE}" ]; then
+	# SYSTEM_PACKAGE can be empty if, for example, it corresponds
+	# to a python package and --enable-system-site-packages was
+	# not passed to ./configure.
```
This script is not invoked within a Sage environment -- it would be better to refer to the environment variable that controls it.

orlitzky replied:
fixed

mkoeppe commented on build/bin/write-dockerfile.sh:
```diff
@@ -15,17 +15,14 @@ SAGE_ROOT=.
+	# to a python package and --enable-system-site-packages was
+	# not passed to ./configure.
+	SYSTEM_PACKAGES+=" ${SYSTEM_PACKAGE}"
+	CONFIGURE_ARGS+="--with-system-${SPKG}=${WITH_SYSTEM_SPKG} "
```
It cannot be assumed that a package with system package information has an spkg-configure and thus that the `--with-system-...` flag exists

orlitzky replied:
fixed

mkoeppe replied:
Thanks. Not 100% the right logic, but good enough.

mkoeppe commented on build/bin/sage-get-system-packages:
```diff
@@ -10,6 +11,12 @@ SPKGS="$*"
+# We source sage-build-env-config because we need to know if
+# --enable-system-site-packages was passed to ./configure.
+# (If not, we want to hide those packages.)
+. "${SAGE_ROOT}/build/bin/sage-build-env-config"
```
That's not good; this script must also work without error in an unconfigured source tree!

mkoeppe replied:
Better to just assume that `ENABLE_SYSTEM_SITE_PACKAGES` is already set in the environment.

orlitzky replied:
I changed this but I'm not sure it makes complete sense. For example `write-dockerfile.sh` now always omits the python packages unless you pass it `ENABLE_SYSTEM_SITE_PACKAGES=yes` manually. That itself is not a big deal but who knows if some other issue is lurking.

mkoeppe replied:
Thanks. Here's an update for tox.ini, which invokes the write-dockerfile.sh
```
--- a/tox.ini
+++ b/tox.ini
@@ -187,6 +187,8 @@ setenv =
     develop:                 EXTRA_SAGE_PACKAGES_4=_develop $(head -n 1 build/pkgs/_develop/dependencies)
     minimal:               SAGE_PACKAGE_LIST_ARGS=_prereq
     maximal:               SAGE_PACKAGE_LIST_ARGS=:standard: :optional:
+    sitepackages:              ENABLE_SYSTEM_SITE_PACKAGES=yes
+    sitepackages:                CONFIG_CONFIGURE_ARGS_SITEPACKAGES=--enable-system-site-packages
     conda-environment:     SAGE_PACKAGE_LIST_ARGS=_prereq
     # Whether to add the system packages needed for bootstrapping
                            EXTRA_SAGE_PACKAGES_0=_bootstrap
@@ -601,7 +603,7 @@ setenv =
     #
     # Resulting full configuration args, including EXTRA_CONFIGURE_ARGS from the user environment
     #
-                      CONFIGURE_ARGS=--enable-experimental-packages --enable-download-from-upstream-url {env:CONFIG_CONFIGURE_ARGS_ROOT:} {env:CONFIG_CONFIGURE_ARGS_1:} {env:CONFIG_CONFIGURE_ARGS_2:} {env:EXTRA_CONFIGURE_ARGS:}
+                      CONFIGURE_ARGS=--enable-experimental-packages --enable-download-from-upstream-url {env:CONFIG_CONFIGURE_ARGS_ROOT:} {env:CONFIG_CONFIGURE_ARGS_SITEPACKAGES:} {env:CONFIG_CONFIGURE_ARGS_1:} {env:CONFIG_CONFIGURE_ARGS_2:} {env:EXTRA_CONFIGURE_ARGS:}
     #
     # Resulting EXTRA_SAGE_PACKAGES
     #
```

orlitzky replied:
You should be able to push to my branch if you want to add that to the repo, dima was able to.

mkoeppe replied:
OK, will do.

mkoeppe replied:
done in e9c4710791

dimpase committed:
lb version in  install-requires.txt for flit

orlitzky committed:
build/bin/write-dockerfile.sh: don't mention --enable-system-site-packages

Since this script can be invoked outside of a sage environment, we
mention the environment variable ENABLE_SYSTEM_SITE_PACKAGES instead.

orlitzky committed:
build/bin/write-dockerfile.sh: skip packages without spkg-configure.m4

We can't add --with-system-foo if foo doesn't have an
spkg-configure.m4 file.

orlitzky committed:
build/bin/sage-get-system-packages: don't source build-env-config

Since sage-get-system-packages is expected to work before ./configure
has been run, we instead rely on ENABLE_SYSTEM_SITE_PACKAGES having
already been set in the environment. When ./configure has not been
run, that variable will be empty, and the python packages will be
omitted from the list.

mkoeppe commented on build/bin/sage-get-system-packages:
```diff
@@ -44,6 +46,17 @@ case "$SYSTEM" in
+    SPKG_CONFIGURE="${SAGE_ROOT}/build/pkgs/${PKG_BASE}/spkg-configure.m4"
+    if grep -q SAGE_PYTHON_PACKAGE_CHECK "${SPKG_CONFIGURE}" 2>/dev/null; then
+	if [ -z "${ENABLE_SYSTEM_SITE_PACKAGES}" ]; then
+	    continue;
```
please untabify

mkoeppe commented on build/bin/sage-get-system-packages:
```diff
@@ -44,6 +46,17 @@ case "$SYSTEM" in
+    # to ./configure (or if ./configure has not yet been run).
+    SPKG_CONFIGURE="${SAGE_ROOT}/build/pkgs/${PKG_BASE}/spkg-configure.m4"
+    if grep -q SAGE_PYTHON_PACKAGE_CHECK "${SPKG_CONFIGURE}" 2>/dev/null; then
+	if [ -z "${ENABLE_SYSTEM_SITE_PACKAGES}" ]; then
```
swap these two `if`s?

orlitzky committed:
build/bin/sage-get-system-packages: swap two "if" statements

It's faster to read a variable than it is to launch grep, so we should
do the easy one first (in case the hard one can be skipped).

orlitzky committed:
build/bin/sage-spkg-info: mention --enable-system-site-packages

Instead of saying that Python packages will never be used, we update
the output of this script to suggest --enable-system-site-packages
when a Python package is queried (and it has an spkg-configure.m4).

mkoeppe commented on build/bin/sage-spkg-info:
```diff
@@ -111,16 +111,17 @@ if [ -z "$system" ]; then
-        echo "If the system package is installed, ./configure will check whether it can be used."
-    else
-        echo "However, these system packages will not be used for building Sage"
         if [ -f "$PKG_SCRIPTS"/install-requires.txt ]; then
```
This is not the right test because of packages like `tox`
(and, hopefully soon, notebook / jupyterlab)

mkoeppe replied:
Renaming the new kind of spkg-configure to `spkg-configure-sitepackage.m4` or something like that might simplify tests so that `grep` is no longer necessary to be invoked

orlitzky replied:
I thought about that or about adding some other piece of metadata like a new package type, but I think grep is the relatively "correct" option. The tangible difference between a "system site package" and any other package right now is the line,

```m4
AS_IF([test "${enable_system_site_packages}" = "yes"], [
```

in `sage_python_package_check.m4`. So in that sense, grepping for `SAGE_PYTHON_PACKAGE_CHECK` is the closest thing we have to a true indicator of whether or not some package needs `--enable-sytem-site-packages`.

mkoeppe replied:
Sure, but the issue is that `sage -package list` now can no longer be used to distinguish between all the relevant package "types"

orlitzky replied:
I don't see any difference between a system python package and a system C (or whatever) package. The experimental option is there right now because it's going to turn on 100+ lightly tested system packages all at once, but the idea of using a python package from the system is not experimental. It used to be possible in sage, and is now a lot safer because we have `install-requires.txt` limiting their use to known-good versions. Once our doctests are updated, install-requires.txt are tightened, and there's a CI job for them, there will be no reason to treat them as a separate class of package.

mkoeppe replied:
I disagree; I think `--enable-system-site-packages` is an option for a separate use case because it is incompatible with contemporary Python development practices -- until distributions find a proper way to support isolated venvs and build isolation.

orlitzky replied:
> until distributions find a proper way to support isolated venvs and build isolation

Yeah nobody wants that.

mkoeppe replied:
> It used to be possible in sage

When? Not even system Python itself could be used before Sage 9.1

mkoeppe replied:
> > until distributions find a proper way to support isolated venvs and build isolation
> 
> Yeah nobody wants that.

I know; but that's part of why distro Python is only relevant as a deployment target but not as a development environment.

mkoeppe replied:
Anyway, this difference of opinion has no bearing on the present PR.

orlitzky replied:
It's the usual distinction between what's easy for developers and what's easy for users. Developers would love to completely ignore the issues of consistent versioning, API/ABI changes, security updates, disk/bandwidth waste, etc. And virtual environments and version pinning largely allow them to do that.

Users, on the other hand, want exactly the opposite. They want the versions to play nicely together, API/ABI upgrades to go smoothly, and all of their software to be patched and up to date without wasting any of their money. Since there are potentially thousands of users for every developer I'm of the opinion that the user should take preference. But regardless, all linux distributions exist solely to cater to their users. It's our job to help them achieve those goals. That's why I say none of us want to isolate anything. It's counterproductive for what we're trying to achieve.

> Anyway, this difference of opinion has no bearing on the present PR.

It kind of does. I see `--enable-system-site-packages` as a stopgap with the goal of eventually treating python packages like any other package. Converting hundreds of python packages to formal second-class citizens would make that a lot harder.

orlitzky replied:
> > It used to be possible in sage
> 
> When? Not even system Python itself could be used before Sage 9.1

Before the virtualenv. It wasn't something that happened automatically but you also didn't have to patch sage to make it possible.

mkoeppe replied:
Well, we'll have to see how usable/maintainable  `--enable-system-site-packages` is going to be; which is why I support merging this PR. I don't insist on a cleaner implementation (that does not fake the package filtering in `sage-get-system-packages`) for this PR.

orlitzky replied:
Ok, thank you. I still have to take a shot at the `./configure` output when I can sit down and stare at it for a while. If the argument to `SAGE_SPKG_CONFIGURE` is always the name of the directory that `spkg-configure.m4` lives in, that may avoid the chicken-and-egg problem.

mkoeppe commented on build/bin/sage-spkg-info:
```diff
@@ -111,16 +111,17 @@ if [ -z "$system" ]; then
-            echo "because using Python site-packages is not supported by the Sage distribution;"
-            echo "see https://github.com/sagemath/sage/issues/29023"
+            echo "If the system package is installed and if --enable-system-site-packages"
+            echo "is passed to ./configure, then ./configure will check if it can be used."
```
Please add "This is experimental." or something like that

orlitzky committed:
build/bin/sage-spkg-info: improve test for system site packages

To avoid false positives with packages like tox (which has an
install-requires.txt but can also be detected without passing
--enable-system-site-packages), we update the test in this script to
grep spkg-configure.m4 rather than simply rely on the presence of both
spkg-configure.m4 and install-requires.txt.

mkoeppe committed:
tox.ini: Add packages factor 'sitepackages'

mkoeppe committed:
build/pkgs/mathjax/distros/gentoo.txt: Shell-quote

mkoeppe committed:
Merge branch 'develop' into scipy_1_11

orlitzky committed:
m4/sage_spkg_configure.m4: reduce python package ./configure noise

When --enable-system-site-packages is not given, the associated
./configure checks short-circuit and will not detect any system
packages. During this no-op it is however rather vociferous:

  ## ---------------------------------------------------------- ##
  ## Checking whether SageMath should install SPKG alabaster... ##
  ## ---------------------------------------------------------- ##
  configure:18401: checking if --enable-system-site-packages was used
  configure:18457: result: no; skipping check
  configure:18487: no suitable system package found for SPKG alabaster

This commit redirects the output from ./configure to hide those
messages in the absence of --enable-system-site-packages. They are
hidden only from the console; config.log retains everything for
debugging purposes.

mkoeppe commented on m4/sage_spkg_configure.m4:
```diff
@@ -58,6 +58,31 @@ dnl   indicate why
+
 echo "-----------------------------------------------------------------------------" >& AS_MESSAGE_FD
 echo "Checking whether SageMath should install SPKG $1..." >& AS_MESSAGE_FD
 AS_BOX([Checking whether SageMath should install SPKG $1...]) >& AS_MESSAGE_LOG_FD
```
wouldn't wrapping these 3 lines by `AS_IF` have been enough?

orlitzky replied:
I wasn't 100% sure how quiet to make it, but I guessed that we should also hide the output from e.g.

```m4
dnl Run DEPS
$6

dnl Run PRE
$4
```

otherwise you'll still see

```
 configure:18401: checking if --enable-system-site-packages was used
 configure:18457: result: no; skipping check
 configure:18487: no suitable system package found for SPKG foo
```

for each python package.

mkoeppe replied:
Oh OK. Fine as is then

mkoeppe committed:
tox.ini (sitepackages): Use IGNORE_MISSING_SYSTEM_PACKAGES=yes

orlitzky committed:
build/pkgs/setuptools_scm/distros/gentoo.txt: fix package name

This was renamed from setuptools_scm to setuptools-scm in Gentoo
on Fri Mar 3 2023.

mkoeppe committed:
.tox, .github/workflows/ci-linux.yml: Test -standard-sitepackages

mkoeppe committed:
build/pkgs: Add distros/*.txt for some Python packages

mkoeppe committed:
build/pkgs: Add distros/*.txt for some more Python packages

mkoeppe committed:
.tox, .github/workflows/ci-linux.yml: Test -standard-sitepackages (fixup)

mkoeppe committed:
build/pkgs/*/distros/debian.txt: Use python3-* packages

mkoeppe committed:
build/pkgs/ipython/distros/debian.txt: Fixup

mkoeppe committed:
.github/workflows/docker.yml: Add opensuse-tumbleweed (currently using python3.11)

mkoeppe committed:
build/pkgs/*/distros/opensuse.txt: Use python3${MINOR}-* packages

mkoeppe committed:
build/bin/write-dockerfile.sh: Substitute ${PYTHON_MINOR}

orlitzky committed:
configure.ac: remove pyvenv.cfg when ./configure is run

This is necessary to trigger a rebuild of pyvenv.cfg, in case the
value of --enable-system-site-packages has changed.

orlitzky committed:
Revert "configure.ac: remove pyvenv.cfg when ./configure is run"

This reverts commit 409ac5ac2135e2a6d912b916f14981d186ae319c.

mkoeppe committed:
build/make/Makefile.in: Use a file separate from pyvenv.cfg as installation record

mkoeppe committed:
build/pkgs/setuptools_scm/distros: Add more

orlitzky committed:
build/make/Makefile.in: ensure $(SAGE_VENV)/$(SPKG_INST_RELDIR) exists

We touch a file inside $(SAGE_VENV)/$(SPKG_INST_RELDIR) to create the
installation record "inst_python3_venv" but that directory must exist
first. We add the prerequisite and "mkdir -p" to create it when it
does not.

mkoeppe committed:
build/pkgs/jupyter_core/dependencies: Add build dep hatchling

mkoeppe committed:
build/pkgs/{pip,setuptools_scm}: Update lower bounds

mkoeppe committed:
Merge branch 'mathjax_gentoo' into system-python-packages

mkoeppe committed:
build/pkgs/setuptools/install-requires.txt: Set upper bound

mkoeppe committed:
build/pkgs/pybind11/install-requires.txt: Increase lower bound

mkoeppe committed:
build/pkgs/pygments/install-requires.txt: Increase lower bound

mkoeppe committed:
build/pkgs/ipython/install-requires.txt: Lower the upper bound

mkoeppe committed:
build/pkgs/cppy/install-requires.txt: Set lower bound

orlitzky committed:
build/pkgs/babel/install-requires.txt: increase lower bound

Require at least babel-2.11.0 for the bugfix,

  * Use email.Message for pofile header parsing instead of the deprecated
    cgi.parse_header function. :gh:`876` - Aarni Koskela

This avoids deprecation warnings that cause some of our doctests to
fail; for example,

  sage: sage_getdoc(foo)
  DeprecationWarning: 'cgi' is deprecated and slated for removal in Python
  3.13
  'Construct a Foo.\n'

in src/doc/en/developer/coding_basics.rst.

orlitzky committed:
build/pkgs/pythran/install-requires.txt: add lower bound

Require at least pythran-0.12.1 for the bugfix in upstream commit
cd68ac68, "Only use numpy.distutils conditionally." Earlier versions
of pythran can emit deprecation warnings that cause the test suite to
fail; for example,

  sage: cython("cdef class A:\n    cdef public int a")
  DeprecationWarning:
      `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
      of the deprecation of `distutils` itself. It will be removed for
      Python >= 3.12. For older Python versions it will remain present.
      It is recommended to use `setuptools < 60.0` for those Python versions.
      For more details, see:
        https://numpy.org/devdocs/reference/distutils_status_migration.html

mkoeppe committed:
build/pkgs/*/dependencies: Reduce $(PYTHON) to an order-only dependency

mkoeppe committed:
build/bin/sage-build-env-config.in: Set PYTHON_MINOR variable

mkoeppe committed:
build/make/Makefile.in: Re-create venv when PYTHON_MINOR, PYTHON_FOR_VENV, or SAGE_VENV_FLAGS change

mkoeppe committed:
.github/workflows/ci-linux.yml (standard-sitepackages): Only test platforms with usable system python

mkoeppe committed:
build/pkgs/{jinja2,packaging}: Increase lower bounds

mkoeppe committed:
build/pkgs/{markupsafe,sphinxcontrib_serializinghtml}: Increase lower bounds

mkoeppe committed:
build/pkgs/sagelib/dependencies: Add pythran as an order-only dep

orlitzky committed:
build/pkgs/python3/spkg-configure.m4: use "cut" for minor component

Use the "cut" utility to parse the python minor version component
instead of sed -E. Despite what the GNU sed manpage says, the "-E"
flag does not actually appear in any version of POSIX, while "cut"
does. The cut version is also a bit more readable.

mkoeppe committed:
Merge branch 'scipy_1_11' into system-python-packages
